src/richie/apps/search/filter_definitions/base.py

```diff
@@ -0,0 +1,382 @@
+
+    def get_fragment_map(self):
+        """
+        Get the possible query fragments for this choice filter.
+
+        Returns:
+        --------
+            Dict: a dictionary mapping the machine name of a value with its human name.
+                For example:
+                {
+                    "is_new": [{"term": {"is_new": True}}],
+                    "is_repeat": [{"term": {"is_new": False}}],
+                }
+        """
+        raise NotImplementedError()
+
+    def get_form_fields(self):
+        """Choice filters are validated with a MultipleChoiceField."""
+        return {
+            self.name: forms.MultipleChoiceField(
+                required=False, choices=self.get_values().items()
+            )
+        }
+
+    def get_faceted_definitions(self, facets, *args, **kwargs):
+        """
+        Add the counts to the values from the initial definition to make them complete
+        as the frontend expects them.
+        """
+        # Get filter values to derive human names
+        values = self.get_values()
+
+        # for each facet, we derive the value and the count:
+        # eg. for facet key `availability@coming_soon`, the value is `coming_soon`
+        facet_counts_dict = {
+            key.split("@")[1]: facet["doc_count"]
+            for key, facet in facets.items()
+            if "{:s}@".format(self.name) in key  # 6 times faster than startswith
+            and facet["doc_count"] > self.min_doc_count
+        }
+        if self.sorting == self.SORTING_NAME:
+            # Alphabetical ascending sorting
+            facet_counts = sorted(facet_counts_dict.items(), key=itemgetter(0))
+        elif self.sorting == self.SORTING_COUNT:
+            # Sorting by descending facet count
+            facet_counts = sorted(
+                facet_counts_dict.items(), key=itemgetter(1), reverse=True
+            )
+        elif self.sorting == self.SORTING_CONF:
+            # Respect the order set in filter definitions
+            facet_counts = [
+                (name, facet_counts_dict[name])
+                for name in values
+                if name in facet_counts_dict
+            ]
+        else:
+            raise ImproperlyConfigured(
+                'Facet sorting should be one of "conf", "count" or "name"'
+            )
+
+        return {
+            self.name: {
+                # We always need to pass the base definition to the frontend
+                "human_name": self.human_name,
+                "is_drilldown": self.is_drilldown,
+                "name": self.name,
+                "position": self.position,
+                "values": [
+                    {"count": count, "human_name": values[name], "name": name}
+                    for name, count in facet_counts
+                ],
+            }
+        }
+
+
+class NestingWrapper(BaseFilterDefinition):
+    """
+    A filter definition to aggregate children filter definitions under a nested query.
```

Comment:
... This is necessary to use filters and aggregations on values that live in nested fields. Such as any information that is linked not directly to courses but to course runs (a nested field of courses).